src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/ConvertToJsonCommandV2.cs

```diff
@@ -0,0 +1,610 @@
+                {
+                    writer.WriteNullValue();
+                }
+
+                return;
+            }
+
+            // Handle Newtonsoft.Json.Linq.JObject by delegating to JsonConverterJObject
+            if (obj is Newtonsoft.Json.Linq.JObject jObject)
+            {
+                System.Text.Json.JsonSerializer.Serialize(writer, jObject, options);
+                return;
+            }
+
+            // If PSObject wraps a primitive type, serialize the base object directly (no depth increment)
+            if (IsPrimitiveType(obj))
+            {
+                System.Text.Json.JsonSerializer.Serialize(writer, obj, obj.GetType(), options);
+                return;
+            }
+
+            // Check depth limit for complex types only (after primitive check)
+            if (currentDepth > _maxDepth)
+            {
+                WriteDepthExceeded(writer, pso, obj);
+                return;
+            }
+
+            // For dictionaries, collections, and custom objects
+            if (obj is IDictionary dict)
+            {
+                SerializeDictionary(writer, pso, dict, options);
+            }
+            else if (obj is IEnumerable enumerable)
+            {
+                SerializeEnumerable(writer, enumerable, options);
+            }
+            else
+            {
+                // For custom objects, serialize as dictionary with properties
+                SerializeAsObject(writer, pso, options);
+            }
+        }
+
+        private void WriteDepthExceeded(Utf8JsonWriter writer, PSObject pso, object obj)
+        {
+            // Write warning once
+            if (!_warningWritten && _cmdlet is not null)
+            {
+                _warningWritten = true;
+                string warningMessage = string.Format(
+                    System.Globalization.CultureInfo.CurrentCulture,
+                    "Resulting JSON is truncated as serialization has exceeded the set depth of {0}.",
+                    _maxDepth);
+                _cmdlet.WriteWarning(warningMessage);
+            }
+
+            // Convert to string when depth exceeded
+            string stringValue = pso.ImmediateBaseObjectIsEmpty
+                ? LanguagePrimitives.ConvertTo<string>(pso)
+                : LanguagePrimitives.ConvertTo<string>(obj);
+            writer.WriteStringValue(stringValue);
+        }
+
+        private void SerializeEnumerable(Utf8JsonWriter writer, IEnumerable enumerable, JsonSerializerOptions options)
+        {
+            writer.WriteStartArray();
+            foreach (var item in enumerable)
+            {
+                if (item is null)
+                {
+                    writer.WriteNullValue();
+                }
+                else
+                {
+                    var psoItem = PSObject.AsPSObject(item);
+
+                    // Recursive call - Write will handle depth tracking
+                    Write(writer, psoItem, options);
```

Comment:
Should we take into account if source object is raw or PSObject? My concern the wrapping will add extra properties (extended and adapted).